tracer/src/Datadog.Trace/RuntimeMetrics/MemoryMappedCounters.cs

```diff
@@ -0,0 +1,593 @@
+                }
+            }
+            catch
+            {
+                view?.Dispose();
+                file?.Dispose();
+                throw;
+            }
+
+            _file = file;
+            _view = view;
+        }
+
+        [Flags]
+        private enum IpcHeaderFlags : ushort
+        {
+            IPC_FLAG_USES_FLAGS = 0x1,
+            IPC_FLAG_INITIALIZED = 0x2,
+            IPC_FLAG_X86 = 0x4
+        }
+
+        public void Dispose()
+        {
+            if (_view != null)
+            {
+                _view.Dispose();
+                _view = null;
+            }
+
+            if (_file != null)
+            {
+                _file.Dispose();
+                _file = null;
+            }
+        }
+
+        public void Refresh()
+        {
+            _view.Read<IpcControlBlock>(0, out var controlBlock);
+
+            var perf = controlBlock.Perf;
+
+            if (perf.GC.ProcessID != _processId)
+            {
+                throw new InvalidOperationException($"The PID in the IPC control block does not match (expected {_processId}, found {controlBlock.Perf.GC.ProcessID}");
+            }
+
+            _statsd.Gauge(MetricsNames.Gen0HeapSize, perf.GC.GenHeapSize0);
+            _statsd.Gauge(MetricsNames.Gen1HeapSize, perf.GC.GenHeapSize1);
+            _statsd.Gauge(MetricsNames.Gen2HeapSize, perf.GC.GenHeapSize2);
+            _statsd.Gauge(MetricsNames.LohSize, perf.GC.LargeObjSize);
+
+            var contentionCount = perf.LocksAndThreads.Contention;
+
+            if (_lastContentionCount == null)
+            {
+                _lastContentionCount = contentionCount;
+            }
+            else
+            {
+                _statsd.Counter(MetricsNames.ContentionCount, contentionCount - _lastContentionCount.Value);
+                _lastContentionCount = contentionCount;
+            }
+
+            var gen0 = GC.CollectionCount(0);
+            var gen1 = GC.CollectionCount(1);
+            var gen2 = GC.CollectionCount(2);
+
+            if (_previousGen0Count != null)
+            {
+                _statsd.Increment(MetricsNames.Gen0CollectionsCount, gen0 - _previousGen0Count.Value);
+            }
+
+            if (_previousGen1Count != null)
+            {
+                _statsd.Increment(MetricsNames.Gen1CollectionsCount, gen1 - _previousGen1Count.Value);
+            }
+
+            if (_previousGen2Count != null)
+            {
+                _statsd.Increment(MetricsNames.Gen2CollectionsCount, gen2 - _previousGen2Count.Value);
+            }
+
+            _previousGen0Count = gen0;
+            _previousGen1Count = gen1;
+            _previousGen2Count = gen2;
+
+            Log.Debug("Sent the following metrics to the DD agent: {metrics}", GarbageCollectionMetrics);
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        private struct TRICOUNT
+        {
```

Comment:
Yeah... I'm just going to take your word for it on all these structures... but any reason this one _isn't_ readonly?

Reply:
Nope, just forgot

Reply:
Fixed